aten/src/THC/generic/THCTensorKthValue.cu

```diff
@@ -0,0 +1,122 @@
+    inputInfo.sizes[dim] = 1;                                             \
+    kthValueInfo.sizes[dim] = 1;                                          \
+    indicesInfo.sizes[dim] = 1;                                           \
+                                                                          \
+    /* Collapse all other dims */                                         \
+    int collapseInputDim = inputInfo.collapseDims(dim);                   \
+    int collapseKthValueDim = kthValueInfo.collapseDims(dim);             \
+    int collapseIndicesDim = indicesInfo.collapseDims(dim);               \
+                                                                          \
+    int64_t inputSlices = 1;                                              \
+    for (int i = 0; i < inputInfo.dims; ++i) {                            \
+      inputSlices *= inputInfo.sizes[i];                                  \
+    }                                                                     \
+    int64_t kthValueSlices = 1;                                           \
+    for (int i = 0; i < kthValueInfo.dims; ++i) {                         \
+      kthValueSlices *= kthValueInfo.sizes[i];                            \
+    }                                                                     \
+                                                                          \
+    dim3 grid;                                                            \
+    if (!THC_getGridFromTiles(inputSlices, grid)) {                       \
+      THError("Slice to select is too large");                            \
+    }                                                                     \
+                                                                          \
+    dim3 block(std::min(THCRoundUp(sliceSize, (int64_t) 32), (int64_t) 1024)); \
+                                                                          \
+    /* This is used as a template parameter to calculate indices. */      \
+    /* We only specialize it if all collapsed dim sizes are the */        \
+    /* same; otherwise, we use -1 which is the specialization */          \
+    /* parameter for arbitrary dimensions */                              \
+    int allDims = inputInfo.dims;                                         \
+    if (kthValueInfo.dims != allDims || indicesInfo.dims != allDims) {    \
+      allDims = -1;                                                       \
+    }                                                                     \
+                                                                          \
+    RUN_DIM(INDEX_T);
+
+    // Based on required index size, run the algorithm with the
+    // appropriate index type
+    if (THCTensor_canUse32BitIndexMath(state, input) &&
+        THCTensor_canUse32BitIndexMath(state, kthValue) &&
+        THCTensor_canUse32BitIndexMath(state, indices)) {
+      RUN_T(uint32_t);
+    } else {
+      RUN_T(uint64_t);
```

Comment:
Could you try running a test case with large indices? You don't have to add it as a unit test (it would probably cause a lot of GPUs to oom...)